File: lab2/csharp/GF.cs
public class DIVIDEbyZeroException:Exception{
    public DIVIDEbyZeroException() :base("!!! CANT DIVIDE BY 0 !!!"){}
} 


public class GF {
    static public long P = 1234567891;

    private long value;
    private long order;

    public GF() {
        this.value = 0;
        this.order = P;
    }
    public GF(long value, long order) {
        this.value = value;
        this.order = order;
    }


    public long getValue() {
        return value;
    }

    public void setValue(long value) {
        this.value = value;
    }


    private static long Inverse(long a)
    {
        long y = 1;
        long n = P - 2;
        long x = a;
        while(n>0){
            if(n%2 == 1)
                y = (y*x) % P;
            n = n/2;
            x = (x*x) % P;
        }

        return y;
    }

    


    public static GF operator + ( GF gf1, GF gf2) {
        GF temp = new GF();
        temp.setValue( (gf1.getValue() + gf2.getValue())% P);
        return temp;
    }
    public static GF operator -( GF gf1,GF gf2) {
        GF temp = new GF();
        temp.setValue((P + (gf1.getValue() - gf2.getValue())) % P);
        return temp;
    }
    public static GF operator * ( GF gf1, GF gf2) {
        GF temp = new GF();
        temp.setValue( (gf1.getValue() * gf2.getValue())% P);
        return temp;
    }
    public static GF operator / ( GF gf1, GF gf2) {
        GF temp = new GF();
        if( gf2.getValue() == 0){
        throw new DIVIDEbyZeroException();
        }

        temp.setValue( (gf1.getValue() * Inverse(gf2.getValue())) % P);

        return temp;
    }

//Operatory z  intami

    public static GF operator + ( GF gf1, long x) {
        GF temp = new GF();
        temp.setValue( (gf1.getValue() + x)% P);
        return temp;
    }
    public static GF operator -( GF gf1,long x) {
        GF temp = new GF();
        temp.setValue((P + (gf1.getValue() - x)) % P);
        return temp;
    }
    public static GF operator * ( GF gf1, long x) {
        GF temp = new GF();
        temp.setValue( (gf1.getValue() * x)% P);
        return temp;
    }
    public static GF operator / ( GF gf1, long x) {
        GF temp = new GF();
        if( x == 0){
        throw new DIVIDEbyZeroException();
        }

        temp.setValue( (gf1.getValue() * Inverse(x)) % P);

        return temp;
    }


    public GF GFAdd(long x) {
        this.value = (this.value + x) % P;
        return this;
    }
    public GF GFsub(long x) {
        this.value = (P + (this.value + x)) % P;
        return this;
    }
    public GF GFmulti(long x) {
        this.value = (this.value * x) % P;
        return this;
    }
    public GF GFdiv(long x) {
        if(x == 0) {
            throw new DIVIDEbyZeroException();
        }
        this.value = (this.value * Inverse(x)) % P;
        return this;
    }



    //ToInt
    public static long ToLong( GF gf)
    {
        return gf.value;
    }

    public override String ToString() {
        return ("GF(" + this.order + ") = " + this.value );
    }
}

File: lab2/cpp/gf.cpp
#include "gf.h"

int64_t inverse(const int64_t a){
    int64_t y =  1;
    int64_t n = P -2;
    int64_t x = a;
    while (n>0) {
        if(n%2 == 1){
            y = (y*x) % P;
        }
        n = n/2;
        x = (x*x) % P;
    }
    return y %P ;
}

//Konstruktory :

GF::GF() {
    order = P;value = 0;
}

GF::GF(int64_t val)  {
   order = P;value = val % P;
   
    if (val < 0)
    {
        value =  P + (val % P);
    }
    
}

GF::~GF() {}


//Getery:

int64_t GF::getOrder() const
{
    return order;
}

int64_t GF::getValue() const
{
    return this->value;
}

//Setery:

void GF::setValue(int64_t v)
{
    value = v;
}

void GF::setOrder(int64_t ord)
{
    order = ord;
}

//Operatory :

GF GF::operator + (const GF obj) const 
{
    GF result;
    result.setValue((value + obj.getValue()) % P);
    return result;
}


GF GF::operator - (const GF obj) const 
{
    GF result;
    result.setValue((P + ((value - obj.getValue()) % P))%P);
    return result;
}

GF GF::operator * (const GF obj) const 
{
    GF result;
    result.setValue(value * obj.getValue() % P);
    return result;
}

GF GF::operator / (const GF obj) const 
{

    if( obj.getValue() ==0 ){
        throw "\n !!! CANT DIVIDE BY 0! !!! \n ";
    }
    GF result;
    result.setValue(value * inverse(obj.getValue()) % P);
    return result;
}

bool GF::operator == (const GF obj) const 
{
    return value == obj.getValue();
}

bool GF::operator != (const GF obj) const 
{
    return value != obj.getValue();
}
bool GF::operator >= (const GF obj) const 
{
    return value >= obj.getValue();
}
bool GF::operator <= (const GF obj) const 
{
    return value <= obj.getValue();
}
bool GF::operator > (const GF obj) const 
{
    return value > obj.getValue();
}
bool GF::operator < (const GF obj) const 
{
    return value < obj.getValue();
}






//Operatory dla działań z typem liczbowym a nie obiektem:

GF GF::operator + (const int64_t x) const 
{
    GF result;
    result.setValue((value + x) % P);
    return result;
}


GF GF::operator - (const int64_t x) const 
{
    GF result;
    result.setValue(P + ((value - x) % P));
    return result;
}

GF GF::operator * (const int64_t x) const 
{
    GF result;
    result.setValue((value * x) % P);
    return result;
}

GF GF::operator / (const int64_t x) const 
{

    if(x ==0 ){
        throw "\n !!! CANT DIVIDE BY 0! !!! \n ";
    }
    GF result;
    result.setValue(value * inverse(x) % P);
    return result;
}

bool GF::operator == (const int64_t x) const 
{
    return value == x;
}

bool GF::operator != (const int64_t x) const 
{
    return value == x;
}

bool GF::operator >= (const int64_t x) const 
{
    return value >= x;
}

bool GF::operator <= (const int64_t x) const 
{
    return value <= x;
}
bool GF::operator > (const int64_t x) const 
{
    return value > x;
}

bool GF::operator < (const int64_t x) const 
{
    return value > x;
}

GF::operator int64_t(){
    return value;
}

std::ostream& operator << (std::ostream& out, const GF obj)
{
    return out << "GF(" << obj.getOrder() << ") =  " << obj.getValue();
}